platformio.ini

```diff
@@ -16,7 +16,7 @@ board_build.filesystem = littlefs
 framework = arduino
 platform = espressif8266
 board = d1_mini_lite
-board_build.ldscript = eagle.flash.4m3m.ld
+board_build.ldscript = eagle.flash.4m2m.ld
```

Comment:
I fear this change will cause breakage for existing users, since the rolling code and client ID are stored in the filesystem already.

Reply:
I just tried this PR.  I don't know if it is related or not, but when I went to Apple Home it thought that my garage door was open when it was closed.  However things quickly synced up, I "closed" the door from Apple Home and nothing happened (to be expected, door was already closed).  But then opening it from Apple Home worked just fine, as did turning off/on the light and closing the door again after it had opened.

During startup the log file did state that rolling code did not exist and created one. Some of the log below.  Bottom line, I think this is fine.

```
>>> [    180] RATGDO: Starting WiFi connecting in background
>>> [    188] HomeKit: Update the CPU to run at 160MHz
>>> [    193] HomeKit: Starting server
>>> [    197] HomeKit: Using existing accessory ID: AF:AB:65:3F:12:D1
>>> [    203] HomeKit: Found admin pairing with 3AD3752B-B0A1-49F5-8327-044D168639D0, disabling pair setup
>>> [    212] HomeKit: Configuring MDNS
>>> [    215] HomeKit: Init server over
>>> [    219] RATGDO: Setting up pins
>>> [    222] RATGDO: enabling built-in LED
>>> [    226] RATGDO: Setting up comms for secplus2.0 protocol
>>> [    342] RATGDO: id_code doesn't exist. creating...
>>> [    351] RATGDO: writing 00 to file id_code
>>> [    360] RATGDO: id code not found
>>> [    369] RATGDO: writing D2539 to file id_code
>>> [    379] RATGDO: id code D2539
>>> [    386] RATGDO: rolling doesn't exist. creating...
>>> [    393] RATGDO: writing 00 to file rolling
>>> [    401] RATGDO: rolling code 00
>>> [    405] RATGDO: Syncing rolling code counter after reboot...
>>> [    412] RATGDO: ENCODING 00000000 00000000000D2539 00000080
>>> [    444] RATGDO: writing 01 to file rolling
>>> [    452] HomeKit: WiFi connected, ip: 192.168.29.88, mask: 255.255.255.0, gw: 192.168.29.3
>>> [    461] HomeKit: Configuring MDNS
>>> [    466] HomeKit: MDNS begin: Garage Door F9EF36, IP: 192.168.29.88
>>> [    981] RATGDO: ENCODING 00000001 00000004000D2539 0000008B
>>> [   1013] RATGDO: writing 02 to file rolling
>>> [   1523] RATGDO: HTTP server started
>>> [   1527] RATGDO: RATGDO setup completed
>>> [   1531] RATGDO: Starting RATGDO Homekit version 0.6.0
>>> [   1536] RATGDO: �[�?
>>> [   1544] RATGDO: reader completed packet
>>> [   1548] RATGDO: DECODED  000180EA 000000C4BC1A200A 020B908C
>>> [   1555] RATGDO: PACKET(0x1A200A @ 0x180EA) Openings - Openings: [Openings 2818]
>>> [   1562] RATGDO: Support for Openings packet unimplemented. Ignoring.
>>> [  25178] HomeKit: Got new client: local 192.168.29.88:5556, remote 192.168.17.224:64726
>>> [  25190] HomeKit: [Client 1073700452] Pair Verify Step 1/2
>>> [  25682] HomeKit: Free heap: 21552
>>> [  25707] HomeKit: [Client 1073700452] Pair Verify Step 2/2
>>> [  25715] HomeKit: [Client 1073700452] Found pairing with 3AD3752B-B0A1-49F5-8327-044D168639D0
>>> [  25745] HomeKit: Call ge_double_scalarmult_vartime_lowmem in ge_low_mem.c
>>> [  27186] HomeKit: [Client 1073700452] Verification successful, secure session established
>>> [  27195] HomeKit: Free heap: 21736
>>> [  27228] HomeKit: [Client 1073700452] Get Accessories
>>> [  27251] RATGDO: get active: 0
>>> [  27256] RATGDO: get current door state: 0
>>> [  27261] RATGDO: get target door state: 0
>>> [  27270] RATGDO: get obstruction: 0
>>> [  27276] RATGDO: get current door state: 0
>>> [  27281] RATGDO: get target door state: 0
>>> [  27292] RATGDO: get light state: Off
>>> [  27348] HomeKit: [Client 1073700452] Get Characteristics
>>> [  27367] HomeKit: [Client 1073700452] Get Characteristics
>>> [  27402] HomeKit: [Client 1073700452] Update Characteristics
>>> [  27428] HomeKit: [Client 1073700452] Update Characteristics
>>> [  27523] HomeKit: [Client 1073700452] Update Characteristics
>>> [  27547] HomeKit: [Client 1073700452] Update Characteristics
>>> [  27567] HomeKit: [Client 1073700452] Update Characteristics
>>> [  27589] HomeKit: [Client 1073700452] Update Characteristics
>>> [  27609] HomeKit: [Client 1073700452] Update Characteristics
>>> [  27634] HomeKit: [Client 1073700452] Update Characteristics
>>> [  27654] HomeKit: [Client 1073700452] Update Characteristics
>>> [  27675] HomeKit: [Client 1073700452] Update Characteristics
>>> [  27695] HomeKit: [Client 1073700452] Update Characteristics
>>> [  27715] HomeKit: [Client 1073700452] Update Characteristics
>>> [  27736] HomeKit: [Client 1073700452] Get Characteristics
>>> [  27767] HomeKit: [Client 1073700452] Get Characteristics
>>> [  27787] HomeKit: [Client 1073700452] Get Characteristics
>>> [  27806] HomeKit: [Client 1073700452] Get Characteristics
>>> [  27836] HomeKit: [Client 1073700452] Get Characteristics
>>> [  27856] HomeKit: [Client 1073700452] Get Characteristics
>>> [  27876] HomeKit: [Client 1073700452] Get Characteristics
>>> [  27898] HomeKit: [Client 1073700452] Get Characteristics
>>> [  27917] HomeKit: [Client 1073700452] Get Characteristics
>>> [  27943] HomeKit: [Client 1073700452] Get Characteristics
>>> [  27963] HomeKit: [Client 1073700452] Get Characteristics
>>> [  28046] HomeKit: [Client 1073700452] Get Characteristics
>>> [  39286] HomeKit: [Client 1073700452] Update Characteristics
>>> [  39310] HomeKit: [Client 1073700452] Get Characteristics
>>> [  39319] RATGDO: get current door state: 0
>>> [  39323] RATGDO: get target door state: 0
>>> [  39328] RATGDO: get obstruction: 0
>>> [  39332] RATGDO: get light state: Off
>>> [  39366] HomeKit: [Client 1073700452] Get Characteristics
>>> [  39388] HomeKit: [Client 1073700452] Get Characteristics
>>> [  65857] HomeKit: [Client 1073700452] Update Characteristics
>>> [  65865] RATGDO: set door state: 1
>>> [  65869] RATGDO: close door req

>>> [  65877] RATGDO: ENCODING 00000002 00000002000D2539 01010080
>>> [  65907] RATGDO: ENCODING 00000002 00000002000D2539 01000080
>>> [  65939] RATGDO: writing 03 to file rolling
>>> [  72128] RATGDO: reader completed packet
>>> [  72132] RATGDO: DECODED  000180EB 000000C0BC1A200A 0008C09D
>>> [  72139] RATGDO: PACKET(0x1A200A @ 0x180EB) UNKNOWN - Unknown: [09D]
>>> [  72145] RATGDO: Support for UNKNOWN packet unimplemented. Ignoring.
>>> [  87106] HomeKit: [Client 1073700452] Update Characteristics
>>> [  87114] RATGDO: set door state: 0
>>> [  87118] RATGDO: open door req

>>> [  87127] RATGDO: ENCODING 00000003 00000002000D2539 01010180
>>> [  87158] HomeKit: [Client 1073700452] Get Characteristics
>>> [  87171] RATGDO: ENCODING 00000003 00000002000D2539 01000180
>>> [  87206] RATGDO: writing 04 to file rolling
>>> [  88279] RATGDO: reader completed packet
>>> [  88284] RATGDO: DECODED  000180EC 000000C0BC1A200A 4260D481
>>> [  88290] RATGDO: PACKET(0x1A200A @ 0x180EC) Status - Status: [DoorState Opening, Parity 0xD, Obs 1, Lock 0, Light 1]
>>> [  88301] RATGDO: activating door
>>> [  88304] HomeKit: [Client 1073700452] Got characteristic 1.10 change event
>>> [  88312] RATGDO: tgt 0 curr 2
>>> [  88315] HomeKit: [Client 1073700452] Got characteristic 1.12 change event
>>> [  88322] HomeKit: [Client 1073700452] Got characteristic 1.11 change event
>>> [  88329] RATGDO: Light Status On
>>> [  88332] HomeKit: [Client 1073700452] Got characteristic 1.21 change event
>>> [  88339] HomeKit: [Client 1073700452] Got characteristic 1.15 change event
>>> [  88346] HomeKit: [Client 1073700452] Got characteristic 1.14 change event
>>> [  89261] RATGDO: reader completed packet
>>> [  89266] RATGDO: DECODED  000180EC 000000C2BC1A200A 0000E084
>>> [  89272] RATGDO: PACKET(0x1A200A @ 0x180EC) MotorOn - NoData: [Zero: 0x00000000, Parity: 0xE]
>>> [  89281] RATGDO: Support for MotorOn packet unimplemented. Ignoring.
>>> [  94415] RATGDO: reader completed packet
>>> [  94420] RATGDO: DECODED  000180EC 000000C2BC1A200A 0000E084
>>> [  94426] RATGDO: PACKET(0x1A200A @ 0x180EC) MotorOn - NoData: [Zero: 0x00000000, Parity: 0xE]
>>> [  94435] RATGDO: Support for MotorOn packet unimplemented. Ignoring.
>>> [  97503] RATGDO: reader completed packet
>>> [  97508] RATGDO: DECODED  000180EC 000000C2BC1A200A 0000E084
>>> [  97514] RATGDO: PACKET(0x1A200A @ 0x180EC) MotorOn - NoData: [Zero: 0x00000000, Parity: 0xE]
>>> [  97523] RATGDO: Support for MotorOn packet unimplemented. Ignoring.
>>> [ 104491] RATGDO: reader completed packet
>>> [ 104495] RATGDO: DECODED  000180ED 000000C0BC1A200A 42608181
>>> [ 104502] RATGDO: PACKET(0x1A200A @ 0x180ED) Status - Status: [DoorState Open, Parity 0x8, Obs 1, Lock 0, Light 1]
>>> [ 104512] RATGDO: tgt 0 curr 0
>>> [ 104515] HomeKit: [Client 1073700452] Got characteristic 1.12 change event
>>> [ 104522] HomeKit: [Client 1073700452] Got characteristic 1.11 change event
>>> [ 104530] HomeKit: [Client 1073700452] Got characteristic 1.15 change event
>>> [ 104537] HomeKit: [Client 1073700452] Got characteristic 1.14 change event
>>> [ 132409] HomeKit: [Client 1073700452] Update Characteristics
>>> [ 132417] RATGDO: set light: Off
>>> [ 132425] RATGDO: ENCODING 00000004 00000002000D2539 00000081
>>> [ 132458] RATGDO: writing 05 to file rolling
>>> [ 133987] HomeKit: [Client 1073700452] Update Characteristics
>>> [ 133995] RATGDO: set door state: 1
```

Reply:
Maybe I should add that I have no idea what changing from `4m3m` to `4m2m` does!!

Reply:
And also that I did not need to re-pair to Apple Home.

Reply:
> I fear this change will cause breakage for existing users, since the rolling code and client ID are stored in the filesystem already.

Good point. Here's what I gather from the code:
- rolling code is stored in filesystem (LittleFS) [link](https://github.com/ratgdo/homekit-ratgdo/blob/d0da1ea3e58a94d52834851684ced93a50eb55cb/src/comms.cpp#L54C7-L54C7)
- device name and serial are inferred from chip id and MAC respectively [link](https://github.com/ratgdo/homekit-ratgdo/blob/d0da1ea3e58a94d52834851684ced93a50eb55cb/src/homekit.cpp#L52-L54)
- Arduino-HomeKit-ESP8266 library stores pairings and other data in EEPROM storage outside of the filesystem [link](https://github.com/mrthiti/Arduino-HomeKit-ESP8266/blob/3825ef4d1c11c79d242c843c094285cbab15ea57/src/storage.c#L26-L35)
- WiFi configuration is also stored outside FS [arduino docs link](https://arduino-esp8266.readthedocs.io/en/3.1.2/filesystem.html)

That only leaves the rolling code getting clobbered by this change, and it seems (from David's comments) even that is quite easily resolved. To fully mitigate even the rolling code issue, one could devise a way to store it in EEPROM (similar to the other bits above) instead of FS. 

Brandon, please advise on how you'd like me to proceed.

Reply:
> Maybe I should add that I have no idea what changing from `4m3m` to `4m2m` does!!

It changes the flash split between code and filesystem (4m total out of which 3m or 2m are reserved for the filesystem). More details in the Arduino docs [here](https://arduino-esp8266.readthedocs.io/en/3.1.2/filesystem.html). 

The change is relevant because it would clobber any data stored in the filesystem. However, per your testing and the code links I posted, the only thing that seems to be reset is the rolling code counter.

Reply:
> To fully mitigate even the rolling code issue, one could devise a way to store it in EEPROM (similar to the other bits above) instead of FS.

Quick note: LittleFS has built-in wear leveling and the current implementation of writes on every rolling code update, which *might* prove problematic if done over and over. On the upside, trimming LittleFS migtt shave a couple kb from the firmware

The [Arduino libraries ](https://arduino-esp8266.readthedocs.io/en/latest/libraries.html) page recommends the following: 
- [ESP_EEPROM](https://github.com/jwrw/ESP_EEPROM) - This library writes a new copy of your data when you save (commit) it and keeps track of where in the sector the most recent copy is kept using a bitmap. The flash sector only needs to be erased when there is no more space for copies in the flash sector.
- [EEPROM Rotate](https://github.com/xoseperez/eeprom_rotate) - Instead of using a single sector to persist the data from the emulated EEPROM, this library uses a number of sectors to do so: a sector pool.

Reply:
@dkerr64 wrote:
> I don't know if it is related or not, but when I went to Apple Home it thought that my garage door was open when it was closed. However things quickly synced up, I "closed" the door from Apple Home and nothing happened (to be expected, door was already closed). But then opening it from Apple Home worked just fine, as did turning off/on the light and closing the door again after it had opened.
> 
> During startup the log file did state that rolling code did not exist and created one.

This, and your log, actually triggered a realization that led to [this commit](https://github.com/ratgdo/homekit-ratgdo/commit/0795e86b9f4947c8d3a5e372d505fa82a543f8d6). Thank you!

@sstoiana I still don't quite see what the benefit to changing the layout is. Why only make 2 MB available to the filesystem versus the previous 3?

Reply:
@thenewwazoo ah, good question. The non-FS portion of flash needs to be large enough to hold both the running code and the ota update, details [here](https://arduino-esp8266.readthedocs.io/en/latest/ota_updates/readme.html#update-process-memory-view).

I actually tried leaving the flash map the same, but the OTA update would fail with an error about insufficient space.

Reply:
Ah! I was under the mistaken impression that the image to be OTA'd was stored in the filesystem.

Reply:
My opinion would be to leave the rolling code in the flash and not the eeprom.  Both have write cycle issues, but the flash is more durable and has the benefit of littleFS having built in wear leveling.  In general, an eeprom should only be used for data that doesn't change frequently.  

Also, given the firmware is still beta and not really considered feature complete, I think it's ok to add some sort of release note that you're going to have to resync your rolling codes.  It's not ideal, but the fixes are non ideal.

Reply:
And I absolutely want to take advantage of wear levelling by the filesystem when writing rolling code.

Reply:
Anyone looked at how good the littleFS wear leveling algorithm is?   Also remember that when we reduce the size of the file system it will also increase your wear

Reply:
I only glossed over the design doc, but there's a loooooong section in the littlefs design doc talking about wear levelling that I read as basically saying "yeah, we do it".

Reply:
it actually does increase the flash wear.  It reduces the file system size from 3M to 2M, so it's cutting down the amount of space for wear leveling by a third.  I haven't yet tried to calculate the actual lifetime, but it may not be an issue at all.  Someone should do the math.

If there is a separate location for emulated eeprom, then the device ID should go there.  It would be nice to move the device ID from the file system to the eeprom, but I think the only way to move it without regenerating it would be to do an intermediate release that moves it before the release with the OTA since that will repartition the flash.

The only concern is that I somewhat suspect there may still be an issue with the device ID generation based on reports of people flashing their device and only getting one valid command before it goes unresponsive.  I know there was a patch for that, but I don't know if that's just a mitigation or a full fix.   I'm worried about trying this PR and losing my working device ID.  This is also pure speculation.  The device ID generation may be solid.

Reply:
The only value stored in the filesystem is the 32-bit rolling code.  Now I have no idea how wear leveling works, but a 1mb filesystem potentially has 250,000 writes before it needs to reuse the same address (or does a write always write the whole minimum block size of 4kb?).

My point is that for storing a single 32-bit value, allocating anything more than the minimum file size seems unnecessarily wasteful.  And I would be shocked if writing that multiple times a day gets anywhere close to stressing the expected life of the flash.

Reply:
Yeah the rolling code is  probably not an issue, but I haven't done all the research to see what type of flash is used and how well the FS used that space. 

There was discussion at some point about capturing logs to flash.  This will be where we run into a problem.  I would leave as must flash as possible for FS if we want to be able to store anything else including logs.